lib/std/Thread/RwLock.zig

```diff
@@ -246,4 +247,132 @@ pub const DefaultRwLock = struct {
         if ((state & READER_MASK == READER) and (state & IS_WRITING != 0))
             rwl.semaphore.post();
     }
+
+    /// The instantaneous value of the waiting-to-write state, for testing only.
+    fn isWaitingToWrite(rwl: *DefaultRwLock) bool {
+        const state = @atomicLoad(usize, &rwl.state, .SeqCst);
+        return (state & DefaultRwLock.WRITER_MASK) != 0;
+    }
 };
+
+test "DefaultRwLock - internal state" {
+    var rwl = DefaultRwLock{};
+
+    // The following failed prior to the fix for Issue #13163,
+    // where the WRITER flag was subtracted by the lock method.
+
+    try testing.expect(!rwl.isWaitingToWrite());
+    rwl.lock();
+    rwl.unlock();
+    try testing.expect(!rwl.isWaitingToWrite());
+}
+
+test "RwLock - smoke test" {
+    var rwl = RwLock{};
+
+    rwl.lock();
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(!rwl.tryLockShared());
+    rwl.unlock();
+
+    try testing.expect(rwl.tryLock());
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(!rwl.tryLockShared());
+    rwl.unlock();
+
+    rwl.lockShared();
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(rwl.tryLockShared());
+    rwl.unlockShared();
+    rwl.unlockShared();
+
+    try testing.expect(rwl.tryLockShared());
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(rwl.tryLockShared());
+    rwl.unlockShared();
+    rwl.unlockShared();
+
+    rwl.lock();
+    rwl.unlock();
+}
+
+test "RwLock - concurrent access" {
+    if (builtin.single_threaded)
+        return;
+
+    const num_writers: usize = 2;
+    const num_readers: usize = 4;
+    const num_writes: usize = 10000;
+
+    const Runner = struct {
+        const Self = @This();
+
+        rwl: RwLock = .{},
+        writes: usize = 0,
+        reads: std.atomic.Atomic(usize) = std.atomic.Atomic(usize).init(0),
+
+        term1: usize = 0,
+        term2: usize = 0,
+        term_sum: usize = 0,
+
+        fn reader(self: *Self) !void {
+            while (true) {
+                self.rwl.lockShared();
+                defer self.rwl.unlockShared();
+                if (self.writes >= num_writes) break;
+
+                try self.check();
+
+                _ = self.reads.fetchAdd(1, .AcqRel);
+            }
+        }
+
+        fn writer(self: *Self, thread_idx: usize) !void {
+            var prng = std.rand.DefaultPrng.init(thread_idx);
+            var rnd = prng.random();
+
+            while (true) {
+                self.rwl.lock();
+                defer self.rwl.unlock();
+                if (self.writes >= num_writes) break;
+
+                try self.check();
+
+                const term1 = rnd.int(usize);
+                self.term1 = term1;
+                try std.Thread.yield();
```

Comment:
FWIW, this is a hint to schedule something else that happens to be ready on this CPU core. Just wanted to mention this as well.

Reply:
The test will not fail if the yields do nothing. The purpose of the yields is to try to bring out bugs in the implementation that can only be seen when switching thread context while holding a lock.

Reply:
Scheduler randomness can be useful for trying to cover more bug surface area, but I believe it shouldn't be relied upon for correctness (anecdotally more so in CI environments).

Reply:
Agreed, and I was not intentionally relying on the yields for the test to finish, as I said.